### PR TITLE
[UWP/WinUI3] Refactor Image Rendering

### PR DESCRIPTION
**Pull request by jwoo-msft**
# Related Issue
Refactored Image loading.

# Description
Changed image loading from completion callback routine to co_routine.
# Sample Card

If appropriate, please include a link to a card in one of the samples directories that can be used to validate this change. This can be an existing card or a card added with this PR.

# How Verified
Verified with all of the sample and test image jsons.
image
image

### CODE REVIEW

File: source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp [hunk 1]
```diff
@@ -367,314 +371,219 @@ namespace AdaptiveCards::Rendering::Xaml_Rendering
...
+                        // as size is being parsed, we can switch to the UI thread to set the source
+                        co_await wil::resume_foreground(GetDispatcher(strongImageSource));
+                        auto svgImageSource = strongImageSource.as<winrt::SvgImageSource>();
 
-                        m_copyStreamOperations.push_back(copyStreamOperation);
+                        if (!isAutoSize)
+                        {
+                            auto size = co_await sizeParseOp;
+                            svgImageSource.RasterizePixelHeight(size.Height);
+                            svgImageSource.RasterizePixelWidth(size.Width);
```
Comment (anna-dingler) [inline source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp:458]: This works for person styling as well? I remember we used to have an aspect ratio issue

File: source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp [hunk 2]
```diff
@@ -705,9 +614,6 @@ namespace AdaptiveCards::Rendering::Xaml_Rendering
             // If the image hasn't loaded yet
             if (imageFiresOpenEvent)
             {
-                // Collapse the Ellipse while the image loads, so that resizing is not noticeable
```
Comment (anna-dingler) [inline source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp:708]: What was this used for before?

File: source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp [hunk 3]
```diff
@@ -367,314 +371,219 @@ namespace AdaptiveCards::Rendering::Xaml_Rendering
...
         {
-            image.as<winrt::BitmapImage>().CreateOptions(winrt::BitmapCreateOptions::None);
-        }
-
-        auto getStreamOperation = httpClient.GetInputStreamAsync(imageUrl);
-        getStreamOperation.Completed(
-            [this, weakThis = this->get_weak(), imgProperties, image](
-                auto const& operation, auto status) -> void
+            // if image loading is handled by Xaml, we set this to true;
+            auto imageFiresOpenEvent{true};
```
Comment (anna-dingler) [inline source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp:435]: Just confirming my understanding:

`imageFiresOpenEvent` is only used by Xaml handling. We default the value to `true` and set it to `false` if `ResolveToStringAsync` returns a stream. `ResolveToStreamAsync` will simply return a nullptr if we did not resolve the stream

File: source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp [hunk 4]
```diff
@@ -367,314 +371,219 @@ namespace AdaptiveCards::Rendering::Xaml_Rendering
...
-                 imageSourceRef = winrt::make_weak(imageSource.as<winrt::SvgImageSource>()), imgProperties](
-                    auto const& result, auto status) -> void
+                else
                 {
-                    auto strongThis = weakThis.get();
-                    auto strongImageSource = imageSourceRef.get();
-                    auto strongStream = streamRef.get();
-
-                    if (strongThis && strongImageSource && strongStream)
+                    // Otherwise, no resolver...
```
Comment (anna-dingler) [inline source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp:472]: What does this comment refer to? I thought the resolver was used within `ResolveToStream`

File: source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp (hunk 1 above)
Comment (anna-dingler) [reply source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp:458]: Just saw the example screenshots so I assume yes 😄

Comment (jwoo-msft) [reply source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp:458]: yes, it works for person styling as well. please see the provided image in the description.

**Review by anna-dingler — APPROVED** [review]

File: source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp (hunk 2 above)
Comment (jwoo-msft) [reply source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp:708]: It's side effects of using co-routine. we don't need to make it invisible.

File: source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp [hunk 5]
```diff
@@ -289,74 +289,78 @@ namespace AdaptiveCards::Rendering::Xaml_Rendering
...
     }
 
-    template<typename TElement>
-    winrt::ImageSource XamlBuilder::SetImageOnUIElement(winrt::Uri const& imageUrl,
-                                                        winrt::AdaptiveCardResourceResolvers const& resolvers,
-                                                        ImageProperties<TElement> const& imgProperties)
+    winrt::Windows::Foundation::Size XamlBuilder::ParseSizeOfSVGImageFromString(winrt::hstring const& content)
     {
-        bool mustHideElement = true;
+        // Parse the size from the XamlDocument
```
Comment (paulcam206) [inline source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp:294]: just so future us don't wonder if we were supposed to be using `XamlDocument` on the next line ;)

```suggestion
        // Parse the size from the XamlDocument as XML
```

File: source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp [hunk 6]
```diff
@@ -289,74 +289,78 @@ namespace AdaptiveCards::Rendering::Xaml_Rendering
...
-        if (resolvers)
+        if (xmlDoc)
         {
-            auto resolver = resolvers.Get(schemeName);
-            // If we have a resolver
-            if (resolver)
+            auto rootElement = xmlDoc.DocumentElement();
+
+            // Root element must be an SVG
+            if (winrt::operator==(rootElement.NodeName(), L"svg"))
```
Comment (paulcam206) [inline source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp:304]: this feels a bit odd -- we don't normally see explicit invocation of operators...

**Review by paulcam206 — APPROVED** [review]

File: source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp (hunk 6 above)
Comment (jwoo-msft) [reply source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp:304]: thanks; I missed it.

**Pull request merged**